Improve error message when inserting empty geometry elements using WFS-T INSERT action

WFS-T INSERT action fails with error message "InvalidParameterValue?: Cannot perform insert operation: Error inserting feature: Index: 0, Size: 0" when geometry element has no values such as:
```
<gml:geometry>
            <gml:MultiSurface></gml:MultiSurface>
</gml:geometry>
```

Exception shown in log (requires log level `log4j.logger.org.deegree.feature.persistence.sql.SQLFeatureStoreTransaction=TRACE`):
```
Caused by: java.lang.IndexOutOfBoundsException: Index: 0, Size: 0
 	at java.util.LinkedList.checkElementIndex(LinkedList.java:555)
 	at java.util.LinkedList.get(LinkedList.java:476)
 	at org.deegree.geometry.standard.multi.DefaultMultiGeometry.getCoordinateDimension(DefaultMultiGeometry.java:85)
 	at org.deegree.geometry.io.WKBWriter.write(WKBWriter.java:70)
 	at org.deegree.sqldialect.postgis.PostGISGeometryConverter.setParticle(PostGISGeometryConverter.java:134)
 	... 37 more
```

Expected is a improved error message in log statement and better ExceptionMessage

--- a/deegree-core/deegree-core-geometry/src/main/java/org/deegree/geometry/io/WKBWriter.java
+++ b/deegree-core/deegree-core-geometry/src/main/java/org/deegree/geometry/io/WKBWriter.java
@@ -35,25 +35,25 @@
  ----------------------------------------------------------------------------*/
 package org.deegree.geometry.io;
 
-import java.io.IOException;
-import java.io.OutputStream;
-
 import org.deegree.geometry.Geometry;
+import org.deegree.geometry.multi.MultiGeometry;
 import org.deegree.geometry.refs.GeometryReference;
 import org.deegree.geometry.standard.AbstractDefaultGeometry;
 import org.deegree.geometry.standard.primitive.DefaultPoint;
-
 import org.locationtech.jts.io.OutputStreamOutStream;
 import org.locationtech.jts.io.ParseException;
 
+import java.io.IOException;
+import java.io.OutputStream;
+
 /**
  * Writes {@link Geometry} objects encoded as Well-Known Binary (WKB).
- * 
+ *
  * TODO re-implement without delegating to JTS TODO add support for non-SFS geometries (e.g. non-linear curves)
- * 
+ *
  * @author <a href="mailto:schneider@lat-lon.de">Markus Schneider</a>
  * @author last edited by: $Author$
- * 
+ *
  * @version $Revision$, $Date$
  */
 public class WKBWriter {
@@ -61,10 +61,19 @@ public class WKBWriter {
     // TODO remove the need for this object
     private static AbstractDefaultGeometry defaultGeom = new DefaultPoint( null, null, null, new double[] { 0.0, 0.0 } );
 
-    public static byte[] write( Geometry geom )
-                            throws ParseException {
+    /**
+     * Exports the passed geom to WKB.
+     *
+     * @param geom
+     *                         never <code>null</code>
+     * @return the WKB as byte array, may be <code>null</code> if the passed geom is an empty multi geometry
+     */
+    public static byte[] write( Geometry geom ) {
         if ( geom instanceof GeometryReference ) {
             geom = ( (GeometryReference<Geometry>) geom ).getReferencedObject();
+        }
+        if ( isEmptyMultiGeometry( geom ) ) {
+            return null;
         }
         // org.locationtech.jts.io.WKBWriter is not thread safe
         int dim = geom.getCoordinateDimension();
@@ -78,4 +87,10 @@ public class WKBWriter {
         new org.locationtech.jts.io.WKBWriter().write( ( (AbstractDefaultGeometry) geom ).getJTSGeometry(),
                                                          new OutputStreamOutStream( os ) );
     }
+
+    private static boolean isEmptyMultiGeometry( Geometry geom ) {
+        return Geometry.GeometryType.MULTI_GEOMETRY.equals( geom.getGeometryType() )
+               && ( (MultiGeometry) geom ).isEmpty();
+    }
+
 }

--- a/deegree-core/deegree-core-geometry/src/main/java/org/deegree/geometry/standard/multi/DefaultMultiGeometry.java
+++ b/deegree-core/deegree-core-geometry/src/main/java/org/deegree/geometry/standard/multi/DefaultMultiGeometry.java
@@ -43,6 +43,7 @@ import java.util.ListIterator;
 import org.deegree.cs.coordinatesystems.ICRS;
 import org.deegree.geometry.Envelope;
 import org.deegree.geometry.Geometry;
+import org.deegree.geometry.GeometryException;
 import org.deegree.geometry.multi.MultiGeometry;
 import org.deegree.geometry.precision.PrecisionModel;
 import org.deegree.geometry.standard.AbstractDefaultGeometry;
@@ -82,7 +83,9 @@ public class DefaultMultiGeometry<T extends Geometry> extends AbstractDefaultGeo
 
     @Override
     public int getCoordinateDimension() {
-        return members.get( 0 ).getCoordinateDimension();
+        if ( isEmpty() )
+            throw new GeometryException( "MultiGeometry is empty, coordinate dimension can not be calculated." );
+        return get( 0 ).getCoordinateDimension();
     }
 
     @Override
@@ -200,7 +203,8 @@ public class DefaultMultiGeometry<T extends Geometry> extends AbstractDefaultGeo
     @Override
     public Envelope getEnvelope() {
         if ( env == null ) {
-            // TODO NullEnvelope for emtpy aggregates? or throw an exception?
+            if ( isEmpty() )
+                throw new GeometryException( "MultiGeometry is empty, coordinate envelope can not be calculated." );
             env = get( 0 ).getEnvelope();
             for ( Geometry geom : this ) {
                 env = env.merge( geom.getEnvelope() );

--- a/deegree-datastores/deegree-mdstores/deegree-mdstore-ebrim-eo/src/main/java/org/deegree/metadata/persistence/ebrim/eo/EbrimEOMDStoreTransaction.java
+++ b/deegree-datastores/deegree-mdstores/deegree-mdstore-ebrim-eo/src/main/java/org/deegree/metadata/persistence/ebrim/eo/EbrimEOMDStoreTransaction.java
@@ -286,21 +286,15 @@ public class EbrimEOMDStoreTransaction implements MetadataStoreTransaction {
         if ( SlotType._geom.equals( slot.getType() ) ) {
             Geometry geom = (Geometry) extrinsicObject.getGeometrySlotValue( SLOTURN + slot.getName() );
             if ( geom != null ) {
-                try {
-                    byte[] wkt = WKBWriter.write( geom );
-                    StringBuilder sb = new StringBuilder();
-                    if ( useLegacyPredicates ) {
-                        sb.append( "SetSRID(GeomFromWKB(?)," );
-                    } else {
-                        sb.append( "SetSRID(ST_GeomFromWKB(?)," );
-                    }
-                    sb.append( "-1)" );
-                    ir.addPreparedArgument( new SQLIdentifier( slot.getColumn() ), wkt, sb.toString() );
-                } catch ( ParseException e ) {
-                    String msg = "Could not write as WKB " + geom + ": " + e.getMessage();
-                    LOG.debug( msg, e );
-                    throw new IllegalArgumentException();
+                byte[] wkt = WKBWriter.write( geom );
+                StringBuilder sb = new StringBuilder();
+                if ( useLegacyPredicates ) {
+                    sb.append( "SetSRID(GeomFromWKB(?)," );
+                } else {
+                    sb.append( "SetSRID(ST_GeomFromWKB(?)," );
                 }
+                sb.append( "-1)" );
+                ir.addPreparedArgument( new SQLIdentifier( slot.getColumn() ), wkt, sb.toString() );
             }
             return;
         }

